### ISSUE TITLE
New lint: Warn against lifetimes being the same, e.g. `'a: 'b, 'b: 'a`

### ISSUE THREAD

**Issue by steffahn** [rust-lang/rust-clippy#7940]
### What it does

Lint against lifetime arguments that come with bounds making them equal.

For example
````rs
fn foo<'a: 'b, 'b: 'a>(x: &'a (), y: &'b ()) {}
````
could be written
````rs
fn foo<'a>(x: &'a (), y: &'a ()) {}
````

Most notably this lint must also be able to take implied bound into consideration, e.g.
````rs
fn foo<'a: 'b, 'b>(x: &'a &'b ()) {}
````

For some background: I came across [this question on URLO](https://users.rust-lang.org/t/lifetime-problem-with-traits/66953); I'm unhappy that neither the Rust compiler nor clippy would be pointing out the problem I pointed out in my answer, even though it's obvious that two lifetimes in question are only accidentally restricted to be restricted to be equal.

### Categories

- Kind: Not quite sure, possibly `suspicions`, or at least `complexity`.

*What is the advantage of the recommended code over the original code*

Shorter, more clear; catches some cases where you're having unnecessary bounds or misdesigned an API

### Drawbacks

Probably none?

I'm not sure under what conditions the lint should fire in trait impls. I suppose there would be cases where bounds from a trait impl and a trait method can end up restricting two lifetimes to be the same without there being a way to avoid it? (I don't have an example (yet) though...)

**Comment by camsteffen** [comment]
You could also make a circle with 3 or more lifetimes.

```rust
fn foo<'a: 'b, 'b: 'c, 'c: 'a>() {}
```

I'd put this in `complexity`.

**Comment by camsteffen** [comment]
Name idea: `codependent_lifetimes`
